aspnetcore/web-api/action-return-types.md

```diff
@@ -0,0 +1,99 @@
+
+Because there are multiple return types and paths in the action, liberal use of the [[ProducesResponseType]](/dotnet/api/microsoft.aspnetcore.mvc.producesresponsetypeattribute.-ctor) attribute is necessary. This attribute produces more descriptive response details for API help pages generated by tools like [Swagger](/aspnet/core/tutorials/web-api-help-pages-using-swagger). It indicates the known types and HTTP status codes to be returned by the action.
+
+### Synchronous action
+
+Consider the following synchronous action in which there are two possible return types:
+
+[!code-csharp[](../web-api/action-return-types/samples/WebApiSample.Api.Pre21/Controllers/ProductsController.cs?name=snippet_GetById&highlight=8,11)]
+
+In the preceding action, a 404 status code is returned when the product represented by `id` doesn't exist in the underlying data store. The [NotFound](/dotnet/api/system.web.http.apicontroller.notfound) helper method is invoked as a shortcut to `return new NotFoundResult();`. If the product does exist, a `Product` object representing the payload is returned with a 200 status code. The [Ok](/dotnet/api/system.web.http.apicontroller.ok) helper method is invoked as the shorthand form of `return new OkObjectResult(product);`.
+
+### Asynchronous action
+
+Consider the following asynchronous action in which there are two possible return types:
+
+[!code-csharp[](../web-api/action-return-types/samples/WebApiSample.Api.Pre21/Controllers/ProductsController.cs?name=snippet_CreateAsync&highlight=8,13)]
+
+In the preceding action, a 400 status code is returned when model validation fails and the [BadRequest](/dotnet/api/system.web.http.apicontroller.badrequest) helper method is invoked. For example, the following model indicates that requests must provide the `Name` property and a value. Therefore, failure to provide a proper `Name` in the request causes model validation to fail.
+
+[!code-csharp[](../web-api/action-return-types/samples/WebApiSample.DataAccess/Models/Product.cs?name=snippet_ProductClass&highlight=5-6)]
+
+The preceding action's other known return code is a 201, which is generated by the [CreatedAtRoute](/dotnet/api/system.web.http.apicontroller.createdatroute) helper method. In this path, the `Product` object is returned.
+
+## ActionResult\<T> type
+
+ASP.NET Core 2.1 introduces `ActionResult<T>` as a valid return type in Web API controller actions. It enables you to return a type deriving from [ActionResult](/dotnet/api/microsoft.aspnetcore.mvc.actionresult) or return a [specific type](#specific-type). `ActionResult<T>` offers the following benefits over the [IActionResult type](#iactionresult-type):
```

Comment:
~valid~
maybe something like
ASP.NET Core 2.1 introduces  the `ActionResult<T>`  return type or Web API controller actions.